File: proto/test/Test/Laws.hs
{-# LANGUAGE OverloadedStrings #-}
-- | The laws of section 5.5, as properties over generated plans.
module Test.Laws (tests, genKnellFree, genItem) where

import qualified Data.Text as T
import Rue.Proto.Algebra
import Rue.Proto.Model
import Test.Tasty (TestTree, testGroup)
import Test.Tasty.HUnit (testCase, (@?=))
import Test.Tasty.QuickCheck

tests :: TestTree
tests =
  testGroup
    "laws"
    [ testProperty "reverse-reverse: reverse . reverse = id on knell-free plans" $
        forAll genKnellFree $ \items ->
          (reverseItems items >>= reverseItems) === Right items
    , testProperty "reverse-seq: reverse (seq a b) = seq (reverse b) (reverse a)" $
        forAll ((,) <$> genKnellFree <*> genKnellFree) $ \(a, b) ->
          reverseItems (seq_ a b) === ((\rb ra -> seq_ rb ra) <$> reverseItems b <*> reverseItems a)
    , testProperty "reverse-par: reverse (par xs) = par (map reverse xs)" $
        forAll genKnellFree $ \xs ->
          reverseItem (par_ xs) === (Par <$> traverse reverseItem xs)
    , testProperty "reverse-knell: a plan with a knell anywhere refuses to reverse" $
        forAll genWithKnell $ \items ->
          case reverseItems items of
            Left (Refused (KnellItem _)) -> property True
            other -> counterexample ("did not refuse: " <> show other) False
    , testProperty "reverse-from: undoing the first k leaves is the reverse of those leaves" $
        forAll genKnellFree $ \items -> forAll (choose (0, length (leaves items))) $ \k ->
          reverseFrom k items === traverse reverseItem (reverse (take k (leaves items)))
    , testProperty "reversing a step flips its direction and nothing else" $
        forAll genStep $ \s ->
          reverseItem (Step s) === Right (Step s {stepDirection = flipD (stepDirection s)})
    , testCase "numbering counts leaves, not containers" $
        map fst (numbered [Step s1, Par [Step s1, Step s1], Confirm, Commit]) @?= [1, 2, 3, 4, 5]
    ]
  where
    flipD Forward = Inverse
    flipD Inverse = Forward
    s1 = step (op "a" [entry Owned "file:/a"])

-- | A knell-free plan of modest depth.
genKnellFree :: Gen [Item]
genKnellFree = sized $ \n -> few n (genItem n)

genWithKnell :: Gen [Item]
genWithKnell = do
  pre <- genKnellFree
  k <- genStep
  post <- genKnellFree
  pure (pre <> [KnellItem k] <> post)

genItem :: Int -> Gen Item
genItem n
  | n <= 0 = Step <$> genStep
  | otherwise =
      frequency
        [ (6, Step <$> genStep)
        , (1, Par <$> few (n `div` 3) (genItem (n `div` 3)))
        , (1, pure Confirm)
        , (1, Observe <$> genName <*> genName)
        , (1, (\g -> Assert g Nothing LapseRevert) <$> genGuard)
        , (1, (\body -> Repeat (Count 2) "i" body) <$> few (n `div` 3) (genItem (n `div` 3)))
        , (1, (\g t e -> When g Nothing LapseRevert t e) <$> genGuard <*> few (n `div` 3) (genItem (n `div` 3)) <*> few (n `div` 3) (genItem (n `div` 3)))
        ]

genStep :: Gen StepI
genStep = do
  o <- genOp
  d <- elements [Forward, Inverse]
  pure (step o) {stepDirection = d}

genOp :: Gen Op
genOp = do
  i <- genName
  k <- elements [Owned, Region, Modified]
  s <- genName
  pure (op i [entry k ("file:/" <> s)])

genGuard :: Gen Guard
genGuard = guard <$> genName <*> elements [Yes, No, Unknown]

genName :: Gen T.Text
genName = T.pack <$> listOf1 (elements ['a' .. 'f'])

few :: Int -> Gen a -> Gen [a]
few n g = choose (0, min 4 (max 0 n)) >>= \k -> vectorOf k g

File: core/tests/laws.rs
//! The laws of section 5.5, as properties over generated plans (a port of the
//! prototype's Test.Laws with a seeded generator in place of QuickCheck).

mod common;

use common::gen::*;
use common::{knell_op, owned, s};
use rue_core::algebra::*;
use rue_core::model::*;

const ROUNDS: u32 = 300;

#[test]
fn reverse_reverse_is_identity_on_knell_free_plans() {
    let mut rng = Rng(0x1234_5678);
    for _ in 0..ROUNDS {
        let items = gen_knell_free(&mut rng);
        let twice = reverse_items(&items).and_then(|r| reverse_items(&r));
        assert_eq!(twice, Ok(items));
    }
}

#[test]
fn reverse_seq_is_seq_of_reverses_swapped() {
    let mut rng = Rng(0x0BAD_CAFE);
    for _ in 0..ROUNDS {
        let a = gen_knell_free(&mut rng);
        let b = gen_knell_free(&mut rng);
        let lhs = reverse_items(&seq_(&a, &b));
        let rhs = reverse_items(&b).and_then(|rb| reverse_items(&a).map(|ra| seq_(&rb, &ra)));
        assert_eq!(lhs, rhs);
    }
}

#[test]
fn reverse_par_is_par_of_reverses() {
    let mut rng = Rng(0x600D_F00D);
    for _ in 0..ROUNDS {
        let xs = gen_knell_free(&mut rng);
        let lhs = reverse_item(&par_(xs.clone()));
        let rhs = xs
            .iter()
            .map(reverse_item)
            .collect::<Result<Vec<_>, _>>()
            .map(|children| Item::Par { children });
        assert_eq!(lhs, rhs);
    }
}

#[test]
fn a_plan_with_a_knell_anywhere_refuses_to_reverse() {
    let mut rng = Rng(0xDEAD_BEEF);
    for _ in 0..ROUNDS {
        let items = gen_with_knell(&mut rng);
        match reverse_items(&items) {
            Err(Refused { at }) if matches!(*at, Item::Knell(_)) => {}
            other => panic!("did not refuse: {other:?}"),
        }
    }
}

#[test]
fn reverse_from_undoes_the_first_k_leaves_in_reverse() {
    let mut rng = Rng(0x5EED_1E55);
    for _ in 0..ROUNDS {
        let items = gen_knell_free(&mut rng);
        let ls = leaves(&items);
        let k = rng.below(ls.len() as u32 + 1) as usize;
        let expected: Result<Vec<Item>, Refused> =
            ls.iter().take(k).rev().map(|it| reverse_item(it)).collect();
        assert_eq!(reverse_from(k, &items), expected);
    }
}

#[test]
fn reversing_a_step_flips_its_direction_and_nothing_else() {
    let mut rng = Rng(0x0F1E_2D3C);
    for _ in 0..ROUNDS {
        let st = gen_step(&mut rng);
        let mut flipped = st.clone();
        flipped.direction = match st.direction {
            Direction::Forward => Direction::Inverse,
            Direction::Inverse => Direction::Forward,
        };
        assert_eq!(reverse_item(&Item::Step(st)), Ok(Item::Step(flipped)));
    }
}

#[test]
fn numbering_counts_leaves_not_containers() {
    let s1 = || s(owned("a"));
    let items = vec![
        s1(),
        Item::Par {
            children: vec![s1(), s1()],
        },
        Item::Confirm,
        Item::Commit,
    ];
    let ns: Vec<u32> = numbered(&items).into_iter().map(|(n, _)| n).collect();
    assert_eq!(ns, vec![1, 2, 3, 4, 5]);
    assert!(knell_free(&items));
    assert!(!knell_free(&[Item::Repeat {
        form: RepeatForm::Count(1),
        var: "i".into(),
        body: vec![Item::Knell(StepI::new(knell_op()))]
    }]));
}
